core/src/main/c/share/dedup.cpp

```diff
@@ -0,0 +1,644 @@
+    const index_t *base = array;
+    int64_t n = count;
+    while (n > 1) {
+        int64_t half = n / 2;
+        MM_PREFETCH_T0(base + half / 2);
+        MM_PREFETCH_T0(base + half + half / 2);
+        auto diff = compare(searchIndex, base[half].i);
+        base = (diff > 0) ? base + half : base;
+        n -= half;
+    }
+    if (compare(searchIndex, base[0].i) == 0) {
+        return base - array;
+    }
+    if (base - array + 1 < count && compare(searchIndex, base[1].i) == 0) {
+        return base - array + 1;
+    }
+    return -1;
+}
+
+template<typename lambda_diff>
+int64_t merge_dedup_long_index_int_keys(
+        const uint64_t *src,
+        const int64_t src_lo,
+        const int64_t src_hi_incl,
+        index_t *const index,
+        const int64_t index_lo,
+        const int64_t index_hi_incl,
+        index_t *dest_index,
+        const lambda_diff compare
+) {
+    int64_t src_pos = src_lo;
+    int64_t index_pos = index_lo;
+    index_t *dest = dest_index;
+
+    bit_vector used_indexes = {};
+    while (src_pos <= src_hi_incl && index_pos <= index_hi_incl) {
+        // Perform normal merge until the timestamp matches.
+        if (src[src_pos] < index[index_pos].ts) {
+            dest[0].ts = src[src_pos];
+            dest[0].i = src_pos | (1ull << 63);
+            dest++;
+            src_pos++;
+        } else if (src[src_pos] > index[index_pos].ts) {
+            *dest = index[index_pos];
+            dest++;
+            index_pos++;
+        } else {
+            // index_ts == src_ts
+            const uint64_t conflict_ts = src[src_pos];
+            index_t *const conflict_index_start = &index[index_pos];
+
+            // Find end of the conflict in index
+            const auto conflict_end_pos = index_pos +
+                                          branch_free_search_lower<index_t>(&index[index_pos],
+                                                                            index_hi_incl - index_pos + 1,
+                                                                            conflict_ts + 1);
+
+            // binary search for matching index record for every source record
+            const int64_t binary_search_len = conflict_end_pos - index_pos;
+
+            // track all found index records
+            used_indexes.reset(binary_search_len);
+            while (src_pos <= src_hi_incl && src[src_pos] == conflict_ts) {
+                dest[0].ts = conflict_ts;
+                const int64_t matched_index = branch_free_search(src_pos, conflict_index_start, binary_search_len,
+                                                                 compare);
+                if (matched_index > -1) {
+                    used_indexes.set(matched_index);
+                    dest[0].i = conflict_index_start[matched_index].i;
+                } else {
+                    dest[0].i = src_pos | (1ull << 63);
```

Comment:
replace with `-src_pos` ?

Reply:
Cannot change these, there is a convention already for merge index